### PR TITLE
first look

### PR DESCRIPTION
**Pull request by darkdef**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ✔️
| Fixed issues  |

### CODE REVIEW

File: psalm.xml [hunk 1]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0"?>
+<psalm
+    errorLevel="6"
```
Comment (samdark) [inline psalm.xml:3]: How about trying to lower it?

File: src/MessageSource.php [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+final class MessageSource implements MessageReaderInterface
+{
+    private string $path;
+    private array $bindedDomains = [];
+
+    public function __construct(string $path)
+    {
+        if (!file_exists($path) || !is_dir($path)) {
+            throw new \RuntimeException(sprintf('Directory "%s" was not exist', $path));
```
Comment (samdark) [inline src/MessageSource.php:17]:
```suggestion
            throw new \RuntimeException(sprintf('Directory "%s" does not exist.', $path));
```

File: src/MessageSource.php [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        }
+        $this->path = $path;
+    }
+
+    /**
+     * function use first parameter as $n for plural if first value - is integer value
```
Comment (samdark) [inline src/MessageSource.php:23]:
```suggestion
     * We use first parameter as `$n` for plurals if its value type is int.
```

File: src/MessageSource.php [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+declare(strict_types=1);
+
+namespace Yiisoft\Translator\Message\Gettext;
+
+use Yiisoft\Translator\MessageReaderInterface;
+
+final class MessageSource implements MessageReaderInterface
+{
+    private string $path;
+    private array $bindedDomains = [];
```
Comment (samdark) [inline src/MessageSource.php:12]:
```suggestion
    private array $boundDomains = [];
```

File: src/MessageSource.php [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        $n = current($parameters);
+        if ($n === false) {
+            return dgettext($category, $id);
+        }
+        return dngettext($category, $id, $id, $n);
+    }
+
+    private function bindDomain(string $category): void
+    {
+        if (!isset($this->bindedDomains[$category])) {
```
Comment (samdark) [inline src/MessageSource.php:38]:
```suggestion
        if (!isset($this->boundDomains[$category])) {
```

File: src/MessageSource.php [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+            bindtextdomain($category, $this->path);
+        }
+    }
+
+    private function setLocale(string $locale): void
+    {
+        if (!setlocale(LC_ALL, $locale)) {
+            throw new \RuntimeException(sprintf('Locale "%s" cannot be setted', $locale));
```
Comment (samdark) [inline src/MessageSource.php:46]:
```suggestion
            throw new \RuntimeException(sprintf('Locale "%s" cannot be set.', $locale));
```

File: tests/MessageSourceTest.php [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+        ];
+    }
+
+    /**
+     * @dataProvider generateTranslationsData
+     */
+    public function testRead(string $expected, string $id, array $params = []): void
+    {
+        $category = 'messages';
+        $locale = PHP_OS_FAMILY === 'Windows' ? 'en-US' : 'en_US';
```
Comment (samdark) [inline tests/MessageSourceTest.php:34]: Interesting way to check it. We've used `DIRECTORY_SEPARATOR === '\\'` previously.

File: tests/MessageSourceTest.php [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+        $messageSource = new MessageSource(__DIR__ . DIRECTORY_SEPARATOR . 'data' . DIRECTORY_SEPARATOR . 'locale');
```
Comment (samdark) [inline tests/MessageSourceTest.php:35]:
```suggestion
        $messageSource = new MessageSource(__DIR__ . '/data/locale');
```

That works well in any OS.

File: tests/MessageSourceTest.php [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    /**
+     * @dataProvider generateTranslationsData
+     */
+    public function testReadWithoutCodepage(string $expected, string $id, array $params = []): void
+    {
+        $category = 'messages';
+        $locale = PHP_OS_FAMILY === 'Windows' ? 'en-US' : 'en_US';
+        $messageSource = new MessageSource(__DIR__ . DIRECTORY_SEPARATOR . 'data' . DIRECTORY_SEPARATOR . 'locale');
```
Comment (samdark) [inline tests/MessageSourceTest.php:47]:
```suggestion
        $messageSource = new MessageSource(__DIR__ . '/data/locale');
```

File: psalm.xml (hunk 1 above)
Comment (darkdef) [reply psalm.xml]: correct to 4

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
Thank you!

**Pull request merged**